File: frontend/src/api/goldToiletOrders.ts
import { apiDelete, apiGet, apiPut } from "./client";

export interface GoldToiletAccount {
  capital: string;
  cash: string;
  updated_at: string;
}

export interface GoldToiletOrderSheet {
  order_date: string;
  entry_percent: string;
  allocation_percent: string;
  loc_percent: string;
  provider_market_open: string | null;
  provider_open_source: string | null;
  provider_open_observed_at: string | null;
  provider_open_last_checked_at: string | null;
  provider_open_failure_reason: string | null;
  manual_market_open: string | null;
  manual_open_observed_at: string | null;
  effective_market_open: string | null;
  effective_open_source: "yahoo_1m_regular_session" | "manual" | null;
  updated_at: string;
}

export interface GoldToiletCalculation {
  breakout_buy_price: string;
  order_quantity: number;
  loc_buy_price: string;
  allocation_amount: string;
  required_reservation_cash: string;
  cash_warning: boolean;
}

export type GoldToiletOpenStatus = "waiting" | "ready" | "failed";

export interface GoldToiletOrderResponse {
  symbol: "SOXL";
  account: GoldToiletAccount | null;
  sheet: GoldToiletOrderSheet | null;
  calculation: GoldToiletCalculation | null;
  open_status: GoldToiletOpenStatus;
  open_failure_reason: string | null;
}

export function getGoldToiletOrderSheet(orderDate: string, signal?: AbortSignal) {
  return apiGet<GoldToiletOrderResponse>(
    `/api/gold-toilet/order-sheet?order_date=${encodeURIComponent(orderDate)}`,
    signal,
  );
}

export function updateGoldToiletAccount(body: { capital: string; cash: string }) {
  return apiPut<GoldToiletOrderResponse>("/api/gold-toilet/account", body);
}

export function updateGoldToiletOrderSheet(
  orderDate: string,
  body: {
    entry_percent: string;
    allocation_percent: string;
    loc_percent: string;
  },
) {
  return apiPut<GoldToiletOrderResponse>(
    `/api/gold-toilet/order-sheet/${encodeURIComponent(orderDate)}`,
    body,
  );
}

export function setGoldToiletManualOpen(orderDate: string, marketOpen: string) {
  return apiPut<GoldToiletOrderResponse>(
    `/api/gold-toilet/order-sheet/${encodeURIComponent(orderDate)}/manual-open`,
    { market_open: marketOpen },
  );
}

export function clearGoldToiletManualOpen(orderDate: string) {
  return apiDelete<GoldToiletOrderResponse>(
    `/api/gold-toilet/order-sheet/${encodeURIComponent(orderDate)}/manual-open`,
  );
}

File: backend/app/dto/gold_toilet_orders.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field


class GoldToiletAccountUpdateDto(BaseModel):
    capital: Decimal = Field(gt=0)
    cash: Decimal = Field(ge=0)


class GoldToiletAccountDto(GoldToiletAccountUpdateDto):
    updated_at: datetime
    model_config = {"from_attributes": True}


class GoldToiletOrderSheetUpdateDto(BaseModel):
    entry_percent: Decimal = Field(gt=-100)
    allocation_percent: Decimal = Field(gt=0, le=100)
    loc_percent: Decimal = Field(gt=-100)
    model_config = {"extra": "forbid"}


class GoldToiletManualOpenUpdateDto(BaseModel):
    market_open: Decimal = Field(gt=0)


class GoldToiletOrderSheetDto(BaseModel):
    order_date: date
    entry_percent: Decimal
    allocation_percent: Decimal
    loc_percent: Decimal
    provider_market_open: Decimal | None
    provider_open_observed_at: datetime | None
    provider_open_source: str | None
    provider_open_last_checked_at: datetime | None
    provider_open_failure_reason: str | None
    manual_market_open: Decimal | None
    manual_open_observed_at: datetime | None
    effective_market_open: Decimal | None
    effective_open_source: Literal["manual", "yahoo_1d_regular_session"] | None
    updated_at: datetime


class GoldToiletCalculationDto(BaseModel):
    breakout_buy_price: Decimal
    order_quantity: int
    loc_buy_price: Decimal
    allocation_amount: Decimal
    required_reservation_cash: Decimal
    cash_warning: bool
    model_config = {"from_attributes": True}


class GoldToiletOrderResponseDto(BaseModel):
    symbol: Literal["SOXL"] = "SOXL"
    account: GoldToiletAccountDto | None
    sheet: GoldToiletOrderSheetDto | None
    allocation_amount: Decimal | None
    calculation: GoldToiletCalculationDto | None
    open_status: Literal["waiting", "ready", "failed"]
    open_failure_reason: str | None = None
